File: homework1/tom_nguyen/q5_ShortestSubstring.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <climits>
using namespace std;

//uses a sliding window to find the shortest substring containing all required chars in O(n)
int shortestSubstring(const string& s, const string& req) {
    unordered_map<char, int> need;
    for (char c : req) {
        ++need[c];
    }
    unordered_map<char, int> window;
    int have = 0; //number of unique chars in req that are satisfied in the window
    int total = need.size(); //number of unique chars to satisfy
    int minLen = INT_MAX;
    int left = 0;
    for (int right = 0; right < (int)s.size(); ++right) {
        ++window[s[right]];
        if (need.contains(s[right]) && window[s[right]] == need[s[right]]) {
            ++have;
        }
        while (have == total) { //shrink window from left
            minLen = min(minLen, right - left + 1);
            --window[s[left]];
            if (need.contains(s[left]) && window[s[left]] < need[s[left]]) {
                --have;
            }
            ++left;
        }
    }
    return (minLen == INT_MAX) ? -1 : minLen;
}

int main() {
    cout << shortestSubstring("abracadabra", "abc") << endl; // 4
    cout << shortestSubstring("zxycbaabcdwxyzzxwdcbxyzabccbazyx", "zzyzx") << endl; // 10
    cout << shortestSubstring("dog", "god") << endl; // 3

    return 0;
}

//time spent: 20 minutes

File: homework1/deepti_patil/q5_ShortestSubstring.py
from collections import Counter

def shortest_substring_length(s: str, required: str) -> int:
    """
    Return the length of the shortest substring of `s` that contains
    all characters in `required` (including duplicates).

    Example:
        s="abracadabra", required="abc" -> 4 ("brac")

    Time Complexity: O(n)
        - Each pointer (left/right) moves across the string at most once.
    Space Complexity: O(m)
        - m is the number of distinct required characters stored in a map.
    """

    # Count how many of each required character we need
    need = Counter(required)
    need_types = len(need)  # number of distinct chars we must satisfy

    # Sliding window counts for characters currently in the window
    window = {}

    have_types = 0  # how many distinct chars currently satisfy the needed count
    best = float("inf")

    left = 0
    for right, ch in enumerate(s):
        # Expand window by including s[right]
        window[ch] = window.get(ch, 0) + 1

        # If this character is required and we just reached the needed amount,
        # we satisfied one required "type"
        if ch in need and window[ch] == need[ch]:
            have_types += 1

        # While the window contains everything we need, try shrinking it
        while have_types == need_types:
            best = min(best, right - left + 1)

            # Shrink from the left
            left_char = s[left]
            window[left_char] -= 1

            # If we removed a required char and now we have too few of it,
            # the window is no longer valid
            if left_char in need and window[left_char] < need[left_char]:
                have_types -= 1

            left += 1

    # If we never found a valid window, return 0 (or you can return -1 if preferred)
    return 0 if best == float("inf") else best


# Examples
print(shortest_substring_length("abracadabra", "abc"))  # 4
print(shortest_substring_length("zxycbaabcdwxyzzxwdcbxyzabccbazxy", "zzyzx"))  # 10
print(shortest_substring_length("dog", "god"))  # 3

